### PR TITLE
Implement TensorView

### PR DESCRIPTION
**Pull request by primenumber**
Implement class template for tensor.

## Motivation and Context

The number of memory layout of tensor is increasing because we will add support for new device. 
In current master implementation, there are no information about memory layout of tensor in C++ code.
This may cause memory layout confuse and difficult to debug.
This PR implement TensorView class to check memory layout compatibility in compile time.
TensorView also enabled easy access to tensor, indexed access for N-dim tensor.
This PR also implement generalized broadcasting binary operator.

## Description

TensorView is implemented as C++ class template.
TensorView has two template parameters: T and memory_layout.
T is element type.
memory_layout is memory layout information enum.
Thanks to memory_layout, we can check memory layout compatibility in compile time.

TensorView has two member variables , ptr and shape.
ptr is pointer points to buffer of tensor.
In order to reuse buffer, ptr doesn't manage memory lifetime. Memory management is user's task.
shape is tensor shape, its length is equal to tensor dimension.

TensorView has some menber functions, operator(), data(), get_offset(), get_offset_ary().
These functions enable indexed access to tensor.

This PR is Rewrite all operators to use TensorView.
binary and unary operators is using binary_op/unary_op template.
binary_op template class is implement generalized broadcasting.

## How has this been tested?

```shell
$ PYTHONPATH=python/dlk python python/dlk/scripts/generate_project.py -i examples/object_detection/fyolo_quantize_4_v4/minimal_graph_with_shape.pb -o tmp/ -p object_detection -hq -ts
$ cd tmp/object_detection
$ make -j8 lm_x86
$ make clean
$ make -j8 lm_arm
$ make clean
$ make -j8 lm_aarch64
$ make clean
$ make -j8 lm_fpga
```
And run test of `lm_x86`, `lm_arm`, `lm_fpga`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
super nice PR.

@tkng @antonionevado @n-nez Please check 👍

File: dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp [hunk 1]
```diff
@@ -26,20 +26,24 @@ limitations under the License.
 
 namespace {
 
+using kn2row_kernel_t = TensorView<QUANTIZED_PACKED_KERNEL, MemoryLayout::HWNC>;
+
 // kernel format converter
 // ohwi : oc kh kw ic, hwoi: kh kw oc ic
-void quantized_ohwi_to_hwoi(const QUANTIZED_PACKED_KERNEL ohwi[], QUANTIZED_PACKED_KERNEL hwoi[], const struct binary_convolution_parameters& p) {
+void quantized_ohwi_to_hwoi(const kernel_t& ohwi, const kn2row_kernel_t& hwoi, const struct binary_convolution_parameters& p) {
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp]: Is it possible to move this function to `tensor_convert.h`?

Comment (tkng) [reply dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp]: I mean, I'd like to hear your idea. If you think it's better to keep it here, it's OK.

Comment (primenumber) [reply dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp]: Implemented new convert_tensor overload.

**Comment by antonionevado** [comment]
Very good PR, thank you so much!
I am trying to run widerface_v5 on x86 and I get this run time error:
```
lm_x86.elf: ./include/tensor_view.h:112: std::size_t TensorView<T, memory_layout>::get_offset(Ts&& ...) const [with Ts = {long unsigned int&}; T = float; MemoryLayout memory_layout = (MemoryLayout)6; std::size_t = long unsigned int]: Assertion `offsets[i] < shape[i]' failed.
```
Not sure why.

**Comment by tkng** [comment]
Sorry, I skip manual test (now I don't have enough time to do so). But the design itself looks good to me.

File: dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp (hunk 1 above)
Comment (tkng) [reply dlk/python/dlk/templates/src/func/impl/generic/quantized_conv2d_kn2row.cpp]: thanks!

**Comment by primenumber** [comment]
Merged tca_support, but network output comparison failed...

**Comment by primenumber** [comment]
Some bugs fixed, but still failed...

**Comment by primenumber** [comment]
Output comparison in widerface_v5 is succeeded!! :tada: 
But more implementation is needed to run other networks.

**Comment by iizukak** [comment]
@ruimashita Please Review 💬

**Comment by primenumber** [comment]
I added comment for memory layout.

**Comment by iizukak** [comment]
@primenumber 
Thanks. Looks good.

@antonionevado 
You added some comment above. Now this PR can be approved?

**Comment by ruimashita** [comment]
@primenumber Thanks 👍 LGTM

**Pull request merged**